Jira Commit Hint Plugin

ierof opened the pull request:
To use plugin you should configure connection to jira in plugin settings and compose query to jira. After all settings are set, new options will appear in "Commit Templates".  Each task from jira query will be a button.
On push any button the key and description of the task will be added to commit message.

You can use simple query like:
_assignee = currentUser() AND status = "In Progress"_
Or some project specific query.
_assignee = currentUser() and resolution is EMPTY and project = Sandbox ORDER BY updatedDate DESC_

![s1](https://user-images.githubusercontent.com/4930272/29413176-1d740818-8364-11e7-94ce-2b06c52409d1.jpg)
![s2](https://user-images.githubusercontent.com/4930272/29267286-af65b24c-80f0-11e7-9dff-e323a4d54066.png)
![s3](https://user-images.githubusercontent.com/4930272/29267290-b1f42b4c-80f0-11e7-9213-f2e0e98a6aa4.png)

vcpp commented on GitUI/GitUICommands.cs:
```diff
@@ -670,19 +670,18 @@ public bool StartFixupCommitDialog(IWin32Window owner, GitRevision revision)
 
         public bool StartCommitDialog(IWin32Window owner, bool showOnlyWhenChanges)
         {
+            var form = new FormCommit(this);
```
I'd prefer to wrap whole method body into `using` block to avoid manual handling of form lifetime and disposing it.

vcpp commented on GitUI/GitUICommands.cs:
```diff
@@ -670,19 +670,18 @@ public bool StartFixupCommitDialog(IWin32Window owner, GitRevision revision)
             };
 
-            return DoActionOnRepo(owner, true, false, PreCommit, PostCommit, action);
+            return DoActionOnRepo(form, true, false, PreCommit, PostCommit, action);
```
What about overloading `DoActionOnRepo` method to support custom pre- and post-event handler types?
Current owner of type `IWin32Window` is provided as parent for modal dialogs, casting it to the specific class and calling it's methods looks a little hacky.

More graceful way is to change `PreCommit` event handler type and give to subscribers special event args class (derived from `GitUIEventArgs`) with features specific for that event.

vcpp commented on Plugins/JiraCommitTipPlugin/JiraCommitTipPlugin.csproj:
```diff
@@ -0,0 +1,88 @@
+  <ItemGroup>
+    <ProjectReference Include="..\..\GitUI\GitUI.csproj">
+      <Project>{CF5B22E7-230F-4E50-BE88-C4F7023CED2C}</Project>
+      <Name>GitUI</Name>
```
Is it ok to reference to the UI project from plugin bypassing the Api?
Instead of using `FormCommit` directly i'd add an interface with minimum set of public members which should be available for using from plugin, put it to the Api assembly (`GitUIPluginInterfaces`) and implement in `FormCommit` class.

vcpp commented on Plugins/JiraCommitTipPlugin/PasswordSetting.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace JiraCommitPlugin
+{
+    public class PasswordSetting : ISetting
```
What's the point of duplicating _Plugins/GitUIPluginInterfaces/Settings/PasswordSetting.cs_?
It seems that this file can be safely removed and it's usages switched to the common implementation provided by the Api.

vcpp commented on Plugins/GitUIPluginInterfaces/Settings/PasswordSetting.cs:
```diff
@@ -0,0 +1,76 @@
+
+            public override TextBox CreateControl()
+            {
+                return new TextBox { PasswordChar = '*' };
```
Hey, asterisks aren't cool, dots are! Try `UseSystemPasswordChar` :-)

vcpp commented on Plugins/GitUIPluginInterfaces/Settings/PasswordSetting.cs:
```diff
@@ -0,0 +1,76 @@
+                {
+                    if (string.IsNullOrEmpty(s))
+                        return string.Empty;
+                    return s;
```
Why not just `return settings.GetValue(Name, string.Empty, s => s ?? string.Empty);`?
One line instead of seven.

vcpp commented on Plugins/JiraCommitTipPlugin/packages.config:
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Atlassian.SDK" version="2.4.0" targetFramework="net40" />
```
New plugin and its third-party libraries should be added to the installer (_Product.wxs_), otherwise end user won't get them.

ierof replied:
fixed

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+        private JiraNamedEntity filter;
+        private readonly StringSetting url = new StringSetting("Jira URL", @"https://jira.atlassian.com");
+        private readonly StringSetting user = new StringSetting("Jira user", string.Empty);
+        private readonly PasswordSetting password = new PasswordSetting("Jira password");
```
What about paranoic-mode when password for jira account isn't saved to the disc and must be entered on each app run? I'm not very comfortable having password saved in plain text.

ierof replied:
Unfortunately, jira provides auth just by plain password. There is workaround -- you can login with guest account in your jira service and compose personal query, not using `currentUser()`

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+        private readonly StringSetting url = new StringSetting("Jira URL", @"https://jira.atlassian.com");
+        private readonly StringSetting user = new StringSetting("Jira user", string.Empty);
+        private readonly PasswordSetting password = new PasswordSetting("Jira password");
+        private readonly StringSetting filterName = new StringSetting("Filter name", "[Filter should be in your favorites]");
```
Hm.. more user-friendly way is to ask for credentials and then load list of available filters, but it seems we cannot provide additional logic for settings page.

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+            Description = description;
+        }
+
+        public override bool Execute(GitUIBaseEventArgs gitUiCommands)
```
Is there a way to avoid calling this method by hiding plugin from direct execution list?

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+            jira = new Jira(url[Settings], user[Settings], password[Settings]);
+            try
+            {
+                if (jira == null || jira.GetAccessToken() == null)
```
It seems that `jira` cannot be `null` here.

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+            }
+            catch (Exception)
+            {
+                return;
```
`return` does nothing.

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+            commitForm = e.OwnerForm as FormCommit;
+            if (commitForm == null)
+                return;
+            var button = CreateAddInfoBitton();
```
Button

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+                Size = new Size(171, 26),
+                Text = description
+            };
+            button.Click -= AddInfo;
```
This is brand new button, we've never been subscribed to it.

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+
+        private void AddInfo(object sender, EventArgs e)
+        {
+            if (commitForm == null)
```
Sharing the same field for each form isn't very robust. I'm not sure if user can open multiple commit forms from the same app instance..
Why not just send the form to the `CreateAddInfoButton` as argument and capture to subscription?

vcpp commented on Plugins/JiraCommitTipPlugin/JiraComminTipPlugin.cs:
```diff
@@ -0,0 +1,126 @@
+            try
+            {
+                return string.Join(Environment.NewLine,
+                    jira.GetIssuesFromFilter(filter.Name).Select(i => i.Key + " " + i.Summary));
```
Mmm.. so many feature requests may be added here :-)
Customizing message template via settings (`{Key} {Summary}`), using drop-down list to select an issue...

vcpp commented on Plugins/JiraCommitTipPlugin/JiraCommitTipPlugin.csproj:
```diff
@@ -0,0 +1,88 @@
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Xml" />
```
5 references are unused.

vcpp commented on Plugins/JiraCommitTipPlugin/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("JiraCommitPlugin")]
```
JiraCommitPlugin vs JiraCommitTipPlugin across all sources..

vcpp commented:
@KindDragon, I've reviewed this PR, comments have been fixed, so can it be merged now?

ierof commented:
@KindDragon also you can add tag "GE-plugin"

ierof commented:
03.08.2016
Review passed.
Merged current master.

jbialobr commented:
Commit Tip button is a nice thing to have. Why don't do it along the lines of the Build server integration implementation? Here is as I would see this:
1. Add Commit Tip button on the FormCommit.
2. Add a Commit Tip subpage into the Commit settings page.
3. The Commit Tip settings page should contain:
 - Enable Commit Tip integration
 - Commit Tip type
 - custom settings for each type, provided by the type implementation (@vcpp here we are able to fetch existing filters)

ierof commented:
@jbialobr It's a good suggestion to move this plugin on "Build server integration" rails. 
I've started this issue before "Build server integration" was released :) About two years ago.
Unfortunately, this pull request not adopted almost two years. For this reason, I have little motivation to change the code. But I'll try to implement your suggesions.

ierof commented:
@jbialobr I've explored "Build server integration" core and notised that is fit to my feature "as is". I'll need to develop something like "Build server integration". If I have enough time, i will do it in some other pull request

ierof commented:
@KindDragon or someone else! Please, approve mo pull request :) 
There is one small change in `IGitUICommands` except plugin.

KindDragon commented:
Sorry, but we can't merge it before 2.49 release

ierof commented:
@KindDragon ok, no problem :) I'm happy that you're not forget about this ancient request :)

jbialobr commented:
> There is one small change in IGitUICommands except plugin.

 I would not call it a small change. It introduces a Technical Debt based on a false assumption that there can be only one commit form at the time.

ierof commented:
@KindDragon @jbialobr I've rewritten this plugin from direct commit form build-in buttons to CommitTemplateItem usage. So there is no more multiple Commit Form issue.

ierof commented:
@jbialobr @KindDragon finally, what about his pull request? :) It'll be three years soon. This code works successfully for about three years. I use this fork as base for my local GE instance, I merge with master regularly.
 
1. It was reviewed by @vcpp 
2. I've rewrite the way I inject text in commit window. Now it's an option in "Commit template". See [commit 8e9ba72573a1b07b9ea9dc6d97e6aa1b90d4545].(https://github.com/gitextensions/gitextensions/pull/2495/commits/8e9ba72573a1b07b9ea9dc6d97e6aa1b90d4545)2. I've updated description. So the issue [mentioned issue](https://github.com/gitextensions/gitextensions/pull/2495#issuecomment-252206164) fixed.
3. I still think that it is very useful plugin for those who use jira as task tracker.

RussKie commented:
@ierof I'm sorry it has taken that long... 
I will review your PR in the next couple of days.

ierof commented:
@RussKie thanks a lot! 
Better late than never :)

RussKie commented on GitUI/CommandsDialogs/CommitDialog/CommitTemplateItem.cs:
```diff
@@ -110,7 +113,28 @@ private static CommitTemplateItem[] DeserializeCommitTemplates(string serialized
+            if (RegistratedTemplates.ContainsKey(key))
+                RegistratedTemplates.Remove(key);
+        }
+
```
I don't think `CommitTemplateItem` is the right place to host this functionality. `CommitTemplateItem` is a DTO which is currently overloaded with behavioral concerns (such as `LoadFromSettings()`). This isn't good as this is ultimately adds to the tech debt.
What I propose we do instead is create a separate class (let's call it `CommitTemplateManager`) which will host your functionality

```
    public interface ICommitTemplateManager
    {
        IEnumerable<CommitTemplateItem> RegisteredTemplates { get; }
        void Register(string templateName, Func<string> templateText);
        void Unregister(string templateName);
    }

    public sealed class CommitTemplateManager : ICommitTemplateManager
    {
        private static readonly ConcurrentDictionary<string, Func<string>> RegisteredTemplatesDic = new ConcurrentDictionary<string, Func<string>>();


        public IEnumerable<CommitTemplateItem> RegisteredTemplates => RegisteredTemplatesDic.Select(item => new CommitTemplateItem(item.Key, item.Value()));


        public void Register(string templateName, Func<string> templateText)
        {
            if (RegisteredTemplatesDic.ContainsKey(templateName))
            {
                return;
            }
            RegisteredTemplatesDic.TryAdd(templateName, templateText);
        }

        public void Unregister(string templateName)
        {
            Func<string> notUsed;
            RegisteredTemplatesDic.TryRemove(templateName, out notUsed);
        }
    }
```

This allows a clean separation of concerns, You can now test you changes as well (for example have a look at `GitRemoteControllerTests`).
We then can later move the remaining serialisation logic from `CommitTemplateItem` into this new class.

ierof replied:
1. The idea of having a static dictionary seems to me not very good....
2. Concerning the creation in the plugi-n an object of type `CommitTemplateManager`: in which assembly should it be in this case? :)

As I understand the architecture of this application:
- `GitUI` assembly should not be connected to plugins.
- The `IGitUICommands` interface in the `GitUIPluginInterfaces` assembly
- Assembly `GitUIPluginInterfaces` is available for plug-ins
- You can get to the context of `IGitUICommands` via `GitUIBaseEventArgs`

So, I think, it can be done without the interface for the first time.

RussKie commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -2459,21 +2459,29 @@ private void commitTemplatesConfigtoolStripMenuItem_Click(object sender, EventAr
                         AddTemplateCommitMessageToMenu(commitTemplates[i], commitTemplates[i].Name);
                 }
+                if (commitTemplates.Any(i => !i.Name.IsNullOrEmpty()))
+                    commitTemplatesToolStripMenuItem.DropDownItems.Add(new ToolStripSeparator());
```
this is a candidate for optimisation - do the union of registered and deserialised templates
something like (pseudo-code)

```
var list = _commitTemplateManager.Registered.Select(x => new { Template = x, Name = x.Name })
                       .Union(new { Template = null, Name = "separator" })
                       .Union(CommitTemplateItem.LoadFromSettings().Select(....));
foreach (var item in commitTemplates)
{
    if (item.Template == null) {
        commitTemplatesToolStripMenuItem.DropDownItems.Add(new ToolStripSeparator());
        continue;
    }
    AddTemplateCommitMessageToMenu(item.Template, item.Name);
}
```

ierof replied:
fixed

RussKie commented on Plugins/JiraCommitTipPlugin/JiraCommitPlugin.cs:
```diff
@@ -0,0 +1,102 @@
+        private readonly StringSetting url = new StringSetting("Jira URL", @"https://jira.atlassian.com");
+        private readonly StringSetting user = new StringSetting("Jira user", string.Empty);
+        private readonly PasswordSetting password = new PasswordSetting("Jira password", string.Empty);
+        private readonly StringSetting filterName = new StringSetting("Filter name", "[Filter should be in your favorites]");
```
Do any of this strings need to be translated?

RussKie commented on Plugins/GitUIPluginInterfaces/IGitUICommands.cs:
```diff
@@ -156,5 +156,7 @@ public interface IGitUICommands
         bool StartViewPatchDialog();
         bool StartSparseWorkingCopyDialog();
+        void AddCommitTemplate(string key, Func<string> addingText);
+        void RemoveCommitTemplate(string key);
```
I don't believe these definitions or the implementing methods are needed - registered templates are stored in a static dictionary so they are accessible universally.

In the plugin you do:

```
private readonly ICommitTemplateManager  _commitTemplateManager = new CommitTemplateManager();

_commitTemplateManager.Register(...);
```

and in `FormCommit` you call: 

```
private readonly ICommitTemplateManager  _commitTemplateManager = new CommitTemplateManager();

var list = _commitTemplateManager.Registered.Select(x => new { Template = x, Name = x.Name })...
```

ierof replied:
I suppose, I have no permission to use `CommitTemplateManager` class in plig-in, I have access only to `IGitUICommands` from `GitUIPluginInterfaces` assembly

ierof replied:
[issuecomment-322020368](https://github.com/gitextensions/gitextensions/pull/2495#issuecomment-322020368)
> The team is currently reviewing the existing architecture and things are likely to change in the future, however until that happens you can use GitCommands.

So, finally, can I request these changes?

RussKie commented on Plugins/JiraCommitTipPlugin/JiraCommitPlugin.cs:
```diff
@@ -0,0 +1,102 @@
+        private string GetMessageToCommit()
+        {
+            if (filter == null)
+                return "<! Your Jira Commit Tip Plugin not configured !> ";
```
Does this need to be translated?

ierof replied:
Most of other plug-ins don't have translations

RussKie commented:
Could you please explain or show what the user gets using this plugin.
On the screenshot you show it is now available in the commit window templates dropdown, but what happens if the user click on it?
If you could provide more information (here's JIRA, here's configuration, here I click and that's what I get) - this would be highly appreciated.

I am also find the name of the plugin a bit confusing/disconnected from the functionality it provides - functionally the plugin deals with templates.
Some clarity is appreciated.

Thank you

ierof commented:
@RussKie At first, thank you for starting reviewing this PR! :)
I'll try to answer all your questions during the week.

ierof commented:
**The wider description of plug-in usage**

1. Most of us, who use Jira as task tracker, have a lot of tasks assigned to him. 
![s1](https://user-images.githubusercontent.com/4930272/29212505-b16d2d22-7ea7-11e7-92c5-41e8c7078676.jpg)

2. And usually committer should describe the task in commit message and also mention task number in commit message.

3. In Jira you can white some filter for tasks you have in some exact status right now. For example, when you press "Start working on issue" the status become "In Progress"
![s2](https://user-images.githubusercontent.com/4930272/29212628-41e136fa-7ea8-11e7-872d-4277694774fc.jpg)
![s4](https://user-images.githubusercontent.com/4930272/29212667-6d1af824-7ea8-11e7-8a42-080e41cee66b.png)

4. So, you can configure my plug-in like this
![s3](https://user-images.githubusercontent.com/4930272/29212634-45a43800-7ea8-11e7-88d9-ef4d22f16b1b.jpg)

5. And then, in commit window, you can press "Commit Template"->"Jira Commit Tip"
 ![s5](https://user-images.githubusercontent.com/4930272/29212709-9f6a82b8-7ea8-11e7-8e19-77e92006b301.png)

6. And task number and task description will be added to commit message textbox
![s6](https://user-images.githubusercontent.com/4930272/29212734-c25e65aa-7ea8-11e7-802b-74b75df43aac.png)

ierof commented:
@RussKie, finally about plug-in name. In my opinion the name describes the functionality of the plug-in. The plugin gives a tip(hint) which message should be in the commit message.

I hope, I've answered all your questions and fixed all change requests.

RussKie commented:
Yes, I think you're right - I was hasty in my refactor suggestions, I will have another look over the weekend.

Thank you for the detailed explanation. I think the idea is great - if a user could automatically load the description from JIRA rather than copy it all manually it would make the user's life heaps easier.
What I understood from your explanation though - right now the plugin addresses a very specific use - the user has to define a public filter and have only one issue in progress. 
Am I getting it right? Is it possible to have several issues in progress?

In the office we use JIRA as well. I often have multiple issues in progress and those issues may belong to different projects (i.e. repos). How could I use your plugin in my setup?
[Another example](https://github.com/gitextensions/gitextensions/pull/3693#issuecomment-310126674) - a user works on 150 repos... This is obviously an extreme use case, but if we imagine for a moment that the user uses JIRA - this plugin can be a significant productivity booster.

Thank you

RussKie commented:
> The plugin gives a tip(hint) which message should be in the commit message.

For English speaking users it looks a bit odd. You are correct that "tip" can mean "hint" but it is often used in [other meanings](http://www.dictionary.com/browse/tip). In the case of this plugin "hint" would be a better choice.

ierof commented:
> I often have multiple issues in progress and those issues may belong to different projects (i.e. repos). How could I use your plugin in my setup?

If have have multiple issues, if's aslo supported.
If you have several "in progress" tasks, all of them will be pasted in commit message line by line.
I have an idea to add several buttons in Commit Templates -- one button per task, but unfortunatly I don't have enough time.

As I know, you can have different GE settings for different repos, so plug-in can be configured to proper filter for each repo. 

And a few more words about multiple "in progress" tasks. I suppose normally people can not do plenty of tasks for one project as the same time. So someone starts task, marks it "In progress" then commits and stops working on it. It not good to hold plenty of "in progess" tasks :)  But anyway, you can create any filter you want to get proper tasks.

RussKie commented:
> The idea of having a static dictionary seems to me not very good....

Why not?

> Concerning the creation in the plugi-n an object of type CommitTemplateManager: in which assembly should it be in this case? :)
> As I understand the architecture of this application:
> GitUI assembly should not be connected to plugins.
> The IGitUICommands interface in the GitUIPluginInterfaces assembly
> Assembly GitUIPluginInterfaces is available for plug-ins
> You can get to the context of IGitUICommands via GitUIBaseEventArgs

> I suppose, I have no permission to use CommitTemplateManager class in plig-in, I have access only to IGitUICommands from GitUIPluginInterfaces assembly

Yes and no, quite a few other plugins reference `GitCommands` project as well. So that would be a place to put the new class in.
The team is currently reviewing the existing architecture and things are likely to change in the future, however until that happens you can use `GitCommands`. 

> And a few more words about multiple "in progress" tasks. I suppose normally people can not do plenty of tasks for one project as the same time. So someone starts task, marks it "In progress" then commits and stops working on it. It not good to hold plenty of "in progess" tasks :) 

We are entering a territory where we start making assumptions and having opinion how people should or should not be working. The purpose of GE to make the life easier, not to tell users how to do their work.
I act as a consultant on multiple projects and naturally I have several WIP items in my queue - each respective project can track and report on my status.
Even on a same project I can have multiple WIP items - I completed work, I submitted a PR and whilst it is being reviewed I start on a next item. According to your logic I should be closing the previous one, which doesn't make sense....

> But anyway, you can create any filter you want to get proper tasks.

Like I said earlier the current implementation seems to address a specific use case, which probably won't work a number of users (myself included). 
One of the ways to make it more usable is to allow a user to select which JIRA item to populate the commit message from (i.e. everytime you open the commit for it queries JIRA for items).
I would like to use your plugin, I think it is great integration piece. People in organisation I work for will certainly benefit from this integration and GE will have an edge over over git UI apps.

ierof commented:
Finally, 
- I've simplified plug-in settings (query instead of filter),
- I improved "several  in progress tasks". Now it's "Each Task = One Button".
- I have updated description. 

So, is there any blockers for this PR now? 
If any, please point to them.
If there is only ideas how to improve this plug-in, I suppose the PR could be approved ;)   

P.S. As I remember, the [Integrating GitExtensions with Build Services](https://blog.pedropombeiro.com/integrating-gitextensions-with-teamcity/) began like small improvement with some bugs. But from release to release I've become good useful feature for GE.

RussKie commented:
> I improved "several in progress tasks". Now it's "Each Task = One Button".

Is there a screenshot you could share?

ierof commented:
> Is there a screenshot you could share?
There are new screenshots in refreshed description :)

vbjay commented:
Fire off a bug in their repo.

On Mon, Aug 14, 2017, 7:57 AM Irina Erofeeva <notifications@github.com>
wrote:

> *@ierof* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormCommit.cs
> <https://github.com/gitextensions/gitextensions/pull/2495#discussion_r132933175>
> :
>
> >              if (null != commitTemplates)
>              {
>                  for (int i = 0; i < commitTemplates.Length; i++)
>                  {
>                      if (!commitTemplates[i].Name.IsNullOrEmpty())
>                          AddTemplateCommitMessageToMenu(commitTemplates[i], commitTemplates[i].Name);
>                  }
> +                if (commitTemplates.Any(i => !i.Name.IsNullOrEmpty()))
> +                    commitTemplatesToolStripMenuItem.DropDownItems.Add(new ToolStripSeparator());
>
> fixed
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/2495#discussion_r132933175>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhscyky-b6LHfWBq6mQlLWyd88qWqBks5sYDYVgaJpZM4CxiuS>
> .
>

RussKie commented:
You can enhance the plugin further:

![image](https://user-images.githubusercontent.com/4403806/29339100-0848158c-825c-11e7-9f49-337aef5cb49e.png)

You can add an ability to configure the template, by either using JIRA specific variable or creating a simple DSL (e.g. `%Summary%` maps to the summary of an issue, `%IssueNumber%` to the JIRA number etc)
And have a "Test" or "Preview" button - this will execute the query and will render the template from the first JIRA available in a new window.

ierof commented:
@RussKie, It's a great idea. 
I will try to implement your request next week. 

But, as I see, there is no engines to insert "Preview" button to Plugin Settings UI now.

ierof commented:
@RussKie 
There is one more commit with templates and preview option: https://github.com/gitextensions/gitextensions/pull/2495/commits/2c2fac30bd603b7a6f0175dc1aed11b6643698bd

RussKie commented:
Looks good.
Could you please confirm that when a user picks an issue from the dropdown on the commit form they get the data in the template they preconfigured?
As in, if I configure a template to be 

```
{Summary}

#{Key}
{Description}
```

willy commit message will be something like this?

```
Fix the annoying bug

#JIRA-123
Description of the annoying bug. 
Bla-bla-bla-bla
```

RussKie commented on Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs:
```diff
@@ -0,0 +1,157 @@
+        {
+            try
+            {
+                return jira.Issues.GetIssuesFromJqlAsync(query).Result
```
why not make the method `async` and await this call?

ierof replied:
Of cause it can be made `async`. But the reason code look like this now is there was no async in old Atlassian SDK. As you can see, the plugin development was started about three years ago :) So, when I've updated Atlassian SDK, I've done just simple support of that SDK, as it was in older version. 
If I have some more free time, I will rewrite this too.

ierof commented:
> Could you please confirm that when a user picks an issue from the dropdown on the commit form they get the data in the template they preconfigured?

Yes, It will be exactly as you describe.

RussKie commented:
Let me know if you need help with mono

ierof commented:
@RussKie  Thanks for the offer. It seems that I fixed the problem https://github.com/gitextensions/gitextensions/pull/2495/commits/ea1534af8a3cdd85216531928c5fbbe6b8841485 . I do not have a studio mono, and my VS2012 changes the order of the projects to But I can't understand if it is right.
